### PR TITLE
Fix collectible NativeCallable UMThunkEntry lifetime

### PR DESCRIPTION
**Pull request by janvorli**
The UMEntryThunk cache entries created for NativeCallable target methods
for collectible classes were not properly cleaned up at the unload time.
This change fixes that by adding UMEntryThunkCache on LoaderAllocator
and using it for entries belonging to NativeCallable targets on
collectible classes. The cache is created lazily.

### CODE REVIEW

File: src/vm/comdelegate.cpp [hunk 1]
```diff
@@ -1076,8 +1076,20 @@ PCODE COMDelegate::ConvertToCallback(MethodDesc* pMD)
     if (NDirect::MarshalingRequired(pMD, pMD->GetSig(), pMD->GetModule()))
         COMPlusThrow(kNotSupportedException, W("NotSupported_NonBlittableTypes"));
 
-    // Get UMEntryThunk from appdomain thunkcache cache.
-    UMEntryThunk *pUMEntryThunk = GetAppDomain()->GetUMEntryThunkCache()->GetUMEntryThunk(pMD);
+    UMEntryThunkCache* pCache;
+    LoaderAllocator* pLoaderAllocator = pMD->GetLoaderAllocator();
+    if ((pLoaderAllocator != NULL) && pLoaderAllocator->IsCollectible())
```
[the remark this answers is not in the document]
Comment (janvorli) [reply src/vm/comdelegate.cpp]: Good point, the answer is  "never". Fixed that.

File: src/vm/comdelegate.cpp [hunk 2]
```diff
@@ -1076,8 +1076,20 @@ PCODE COMDelegate::ConvertToCallback(MethodDesc* pMD)
     if (NDirect::MarshalingRequired(pMD, pMD->GetSig(), pMD->GetModule()))
         COMPlusThrow(kNotSupportedException, W("NotSupported_NonBlittableTypes"));
 
-    // Get UMEntryThunk from appdomain thunkcache cache.
-    UMEntryThunk *pUMEntryThunk = GetAppDomain()->GetUMEntryThunkCache()->GetUMEntryThunk(pMD);
+    UMEntryThunkCache* pCache;
+    LoaderAllocator* pLoaderAllocator = pMD->GetLoaderAllocator();
+    if (pLoaderAllocator->IsCollectible())
```
Comment (jkotas) [inline src/vm/comdelegate.cpp]: Is there a reason for this check - why can't be just do `GetLoaderAllocator()->GetUMEntryThunkCache()` unconditionally?

Comment (janvorli) [reply src/vm/comdelegate.cpp]: Hmm, interesting. So you mean we would then remove the cache from the AppDomain and have it only in the LoaderAllocator. That sounds fine. Although we could possibly end up with more instances of the cache, the overhead would be negligible.

Comment (jkotas) [reply src/vm/comdelegate.cpp]: Right.

> we could possibly end up with more instances of the cache

I do not think we would end up with more instances of the cache.

**Comment by janvorli** [comment]
@jkotas I've made the changes based on your feedback, can you please take a look again?

**Pull request merged**